components/raftstore/src/store/fsm/apply.rs

```diff
@@ -799,26 +716,34 @@ where
     }
 
     /// Handles all the committed_entries, namely, applies the committed entries.
-    fn handle_raft_committed_entries<W: WriteBatch<EK>>(
+    async fn handle_raft_committed_entries<W: WriteBatch<EK>>(
         &mut self,
         apply_ctx: &mut ApplyContext<EK, W>,
-        mut committed_entries_drainer: Drain<Entry>,
+        committed_entries: Vec<Entry>,
     ) {
-        if committed_entries_drainer.len() == 0 {
-            return;
-        }
         apply_ctx.prepare_for(self);
+        if apply_ctx.timer.is_none() {
+            apply_ctx.timer = Some(Instant::now_coarse());
+        }
         // If we send multiple ConfChange commands, only first one will be proposed correctly,
         // others will be saved as a normal entry with no data, so we must re-propose these
         // commands again.
-        apply_ctx.committed_count += committed_entries_drainer.len();
         let mut results = VecDeque::new();
-        while let Some(entry) = committed_entries_drainer.next() {
+        for entry in committed_entries {
             if self.pending_remove {
                 // This peer is about to be destroyed, skip everything.
                 break;
             }
 
+            if apply_ctx.kv_wb().should_write_to_engine() {
```

Comment:
How about moving this to the end of the loop, it will be more straightforward.

Reply:
No. If this `Apply` msg include a lot of entries, we shall `reschedule` in middle of the loop.

Reply:
I think here just the start of the loop, and there is not much difference between the `start` and the `end` of a loop. In contrast, after handling the last log entries, the writebatch may large enough to trigger a write to the engine, and we can't do that because the check on the start of the loop.